tooling/greybox_fuzzer/src/corpus.rs

```diff
@@ -0,0 +1,431 @@
+        Self {
+            executions_per_testcase: HashMap::new(),
+            sequence_number: HashMap::new(),
+            total_executions: 0,
+            current_sequence: Sequence::new(),
+        }
+    }
+
+    /// Adds a new testcase to be managed by the orchestrator
+    pub fn new_testcase(&mut self, testcase_id: TestCaseId) {
+        self.executions_per_testcase.insert(testcase_id, 0);
+        self.sequence_number.insert(testcase_id, 0);
+    }
+
+    /// Removes a testcase from being managed by the orchestrator.
+    /// Updates execution counts and clears current sequence if it was using this testcase.
+    pub fn remove(&mut self, testcase_id: TestCaseId) {
+        let executions = self.executions_per_testcase[&testcase_id];
+        self.executions_per_testcase.remove(&testcase_id);
+        self.sequence_number.remove(&testcase_id);
+        self.total_executions -= executions;
+        if self.current_sequence.testcase_id == testcase_id {
+            self.current_sequence.clear();
+        }
+    }
+
+    /// Selects the next testcase(s) to use for fuzzing based on execution counts.
+    /// Prioritizes testcases that have been executed least often.
+    /// Returns both a primary testcase and optionally a second one for splicing.
+    pub fn get_next_testcase(&mut self, prng: &mut XorShiftRng) -> NextSelection {
+        let testcase_count = self.executions_per_testcase.len();
+        // If we're in the middle of a sequence, continue using the current testcase
+        if !self.current_sequence.is_empty() {
+            // Update execution counts
+            self.current_sequence.decrement();
+            self.executions_per_testcase
+                .entry(self.current_sequence.testcase_id)
+                .and_modify(|executions| *executions += 1);
+            self.total_executions += 1;
+        } else {
+            // Calculate average executions per testcase
+            let average = self.total_executions / testcase_count as u64;
+
+            // Select the testcase with the lowest execution count
+            let chosen_id = self
+                .executions_per_testcase
+                .iter()
+                .min_by_key(|&(_id, value)| value)
+                .map(|(&id, _)| id)
+                .unwrap();
+
+            // Increment sequence number for chosen testcase
+            self.sequence_number
+                .entry(chosen_id)
+                .and_modify(|sequence_number| *sequence_number += 1);
+
+            // Create new sequence with exponentially increasing length based on sequence number
+            self.current_sequence = Sequence {
+                testcase_id: chosen_id,
+                executions_left: min(1u64 << self.sequence_number[&chosen_id], average / 2),
+            };
+
+            // Update execution counts
+            self.total_executions += 1;
+            self.executions_per_testcase.entry(chosen_id).and_modify(|executions| *executions += 1);
```

Comment:
It seems the first statement could be moved outside of this `if`.

Then, at this point using `chosen_id` or `self.current_sequence.testcase_id` is the same, so this similar line from the other branch could also be moved outside of this `if`.